### PR TITLE
Add gitignore for maven, gradle, java, intellij and eclipse

### PR DESCRIPTION
**Pull request by thechubbypanda**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,9 +1,217 @@
...
+.idea/**/dataSources.ids
+.idea/**/dataSources.local.xml
+.idea/**/sqlDataSources.xml
+.idea/**/dynamic.xml
+.idea/**/uiDesigner.xml
+.idea/**/dbnavigator.xml
+
+# Gradle
+.idea/**/gradle.xml
+.idea/**/libraries
```
Comment (payne911) [inline .gitignore:99]: Not really necessary because of ``L148``.

File: .gitignore [hunk 2]
```diff
@@ -1,9 +1,217 @@
...
+
+# Gradle and Maven with auto-import
+# When using Gradle or Maven with auto-import, you should exclude module files,
+# since they will be recreated, and may cause churn.  Uncomment if using
+# auto-import.
+# .idea/modules.xml
+# .idea/*.iml
+# .idea/modules
+# *.iml
+# *.ipr
```
Comment (payne911) [inline .gitignore:109]: Can be removed (even uncommenting this would have no effect: those are already covered elsewhere).

File: .gitignore [hunk 3]
```diff
@@ -1,9 +1,217 @@
...
+
+# CMake
+cmake-build-*/
+
+# Mongo Explorer plugin
+.idea/**/mongoSettings.xml
```
Comment (payne911) [inline .gitignore:115]: Can be removed. See ``L148``.

File: .gitignore [hunk 4]
```diff
@@ -1,9 +1,217 @@
...
+out/
+
+# mpeltonen/sbt-idea plugin
+.idea_modules/
+
+# JIRA plugin
+atlassian-ide-plugin.xml
+
+# Cursive Clojure plugin
+.idea/replstate.xml
```
Comment (payne911) [inline .gitignore:130]: Can be removed. See ``L148``.

File: .gitignore [hunk 5]
```diff
@@ -1,9 +1,217 @@
...
+com_crashlytics_export_strings.xml
+crashlytics.properties
+crashlytics-build.properties
+fabric.properties
+
+# Editor-based Rest Client
+.idea/httpRequests
+
+# Android studio 3.1+ serialized cache file
+.idea/caches/build_file_checksums.ser
```
Comment (payne911) [inline .gitignore:142]: Can be removed. See ``L148``.

File: .gitignore [hunk 6]
```diff
@@ -1,9 +1,217 @@
...
+
 .idea/
-libs/
-target/
+
+# Reason: https://github.com/joeblau/gitignore.io/issues/186#issuecomment-249601023
+
+*.iml
+modules.xml
+.idea/misc.xml
```
Comment (payne911) [inline .gitignore:154]: Can be removed. See ``L148``.

File: .gitignore [hunk 7]
```diff
@@ -1,9 +1,217 @@
...
+target/
+pom.xml.tag
+pom.xml.releaseBackup
+pom.xml.versionsBackup
+pom.xml.next
+release.properties
+dependency-reduced-pom.xml
+buildNumber.properties
+.mvn/timing.properties
+.mvn/wrapper/maven-wrapper.jar
```
Comment (payne911) [inline .gitignore:195]: ``.jar`` are already ignored (``L174``).

File: .gitignore [hunk 8]
```diff
@@ -1,9 +1,217 @@
...
+
+### Gradle ###
+.gradle
+build/
+
+# Ignore Gradle GUI config
+gradle-app.setting
+
+# Avoid ignoring Gradle wrapper jar file (.jar files are usually ignored)
+!gradle-wrapper.jar
```
Comment (payne911) [inline .gitignore:206]: Contradicts ``L174`` and ``L195``.

File: .gitignore [hunk 9]
```diff
@@ -1,9 +1,217 @@
...
+
+# Cache of project
+.gradletasknamecache
+
+# # Work around https://youtrack.jetbrains.com/issue/IDEA-116898
+# gradle/wrapper/gradle-wrapper.properties
+
+### Gradle Patch ###
+**/build/
```
Comment (payne911) [inline .gitignore:215]: See ``L200``. This should probably be removed (unless ``build`` folders are produced elsewhere than at root?).

File: .gitignore [hunk 10]
```diff
@@ -1,9 +1,217 @@
...
+
+# Reason: https://github.com/joeblau/gitignore.io/issues/186#issuecomment-249601023
+
+*.iml
+modules.xml
+.idea/misc.xml
+*.ipr
+
+# Sonarlint plugin
+.idea/sonarlint
```
Comment (payne911) [inline .gitignore:158]: Can be removed. See ``L148``.

**Comment by mgsx-dev** [comment]
also, gitignore patterns like "out" or "build" are dangerous, see https://github.com/libgdx/libgdx/pull/4580 where we fixed potential false positive.

File: .gitignore (hunk 9 above)
Comment (thechubbypanda) [reply .gitignore:166]:
```suggestion
```

File: .gitignore (hunk 7 above)
Comment (thechubbypanda) [reply .gitignore:146]:
```suggestion
```

File: .gitignore (hunk 8 above)
Comment (thechubbypanda) [reply .gitignore:157]: I actually don't know why this is here, I'm happy to remove it?

**Review by thechubbypanda — COMMENTED** [review]
Removed most of the redundancies

**Comment by thechubbypanda** [comment]
> also, gitignore patterns like "out" or "build" are dangerous, see [libgdx/libgdx#4580](https://github.com/libgdx/libgdx/pull/4580) where we fixed potential false positive.

I'm happy to change them but I don't entirely understand the issue. You're welcome to make a suggested change

**Comment by mgsx-dev** [comment]
@thechubbypanda see https://github.com/libgdx/libgdx/issues/4560 you'll understand :-)
the idea is to ignore absolute paths instead of general patterns for some folders.

**Review by thechubbypanda — COMMENTED** [review]
Committed changes

**Review by thechubbypanda — COMMENTED** [review]
Accounted for https://github.com/libgdx/libgdx/issues/4560

**Comment by mgsx-dev** [comment]
@thechubbypanda nice! i think you forgot the target/ dir to be replaced by /target/ (not sure ./ is a thing with gitignore syntax, simply / is enough iirc)

**Pull request closed without merging**